### PR TITLE
Set fan speed by percent.

### PR DESCRIPTION
**Pull request by MatthewHink**

### CODE REVIEW

**Comment by MatthewHink** [comment]
Kiss my ass on the code coverage report.

File: synse/routes/aliases.py [hunk 1]
```diff
@@ -102,33 +103,76 @@
...
-            'action': 'speed',
-            'raw': param_speed
-        }
-        transaction = await commands.write(rack, board, device, data)
-        return transaction.to_json()
-
-    # if no request parameter is specified, this will translate to a
-    # read request.
+    # This sucks. Get the first key in request.raw_args. Example: request.raw_args: {'speed': '0'}
+    if len(request.raw_args) == 1:
```
[the remark this answers is not in the document]
Comment (MatthewHink) [reply synse/routes/aliases.py]: What would you propose for checking this?

Valid query parameters would be:
<none> for read
speed for rpm
speed_percent for 0-100 percent.

File: synse/routes/aliases.py [hunk 2]
```diff
@@ -102,33 +103,76 @@
...
+                    request, request.raw_args))
+
+            # elif param_speed_percent is not None:
+            # FIXME - is the below true? could we check against the device prototype's
+            #   "range.min" and "range.max" fields for this?
+            # no validation is done on the fan speed. valid fan speeds vary based on
+            # fan make/model, so it is up to the underlying implementation to do the
+            # validation.
+            # ^ mhink - Yes it absolutely is and the underlying code should fail as needed.
+            # Also max and min speeds vary by the fan motor, not the controller. (the device)
```
[the remark this answers is not in the document]
Comment (MatthewHink) [reply synse/routes/aliases.py]: I'll update the comment.

File: synse/routes/aliases.py [hunk 3]
```diff
@@ -102,33 +103,76 @@
...
+            data = {
+                'action': 'speed_percent',
+                'raw': param_speed_percent,
+            }
+            transaction = await commands.write(rack, board, device, data)
+            return transaction.to_json()
+        # If no request parameter is specified, this will translate to a
+        # read request.
+        else:
+            # TODO: This is dodgy.
```
[the remark this answers is not in the document]
Comment (MatthewHink) [reply synse/routes/aliases.py]: I'll pull this out.

[the remark this answers is not in the document]
Comment (MatthewHink) [reply synse/routes/aliases.py]: I may not be able to pull this out if passing in both speed and speed_percent gets past the validation.

**Comment by edaniszewski** [comment]
In the above, it looks like you're using the wrong curl for writing (e.g. the `/read/` portion should be `/fan/`:
```
curl http://${SYNSE_IP}:5000/synse/2.0/read/chamber/vec/961e1b0ae21c404a0a6a7e47224570ec?speed_percent=50
```

**Comment by MatthewHink** [comment]
Well shit. Than that's another bug. Let's roll this back if this is the cause.

```
vapor@vec1-c1-wrigley:~$ curl http://${SYNSE_IP}:5000/synse/2.0/fan/chamber/vec/961e1b0ae21c404a0a6a7e47224570ec
{
  "type":"fan",
  "data":{
    "fan_speed":{
      "value":438.0,
      "timestamp":"2018-04-12T21:19:15.155782397Z",
      "unit":{
        "symbol":"RPM",
        "name":"revolutions per minute"
      }
    },
    "direction":{
      "value":"forward",
      "timestamp":"2018-04-12T21:19:15.155785715Z",
      "unit":null
    },
    "max_rpm":{
      "value":"1755",
      "timestamp":"2018-04-12T21:19:15.155787795Z",
      "unit":null
    },
    "min_rpm":{
      "value":"175",
      "timestamp":"2018-04-12T21:19:15.155789559Z",
      "unit":null
    }
  }
}
vapor@vec1-c1-wrigley:~$ curl http://${SYNSE_IP}:5000/synse/2.0/fan/chamber/vec/961e1b0ae21c404a0a6a7e47224570ec?speed_percent=50
[
  {
    "context":{
      "action":"speed_percent",
      "raw":[
        "50"
      ]
    },
    "transaction":"bb7sqrppr5tg00fh8m80"
  }
]
vapor@vec1-c1-wrigley:~$ curl http://${SYNSE_IP}:5000/synse/2.0/fan/chamber/vec/961e1b0ae21c404a0a6a7e47224570ec
{
  "type":"fan",
  "data":{
    "fan_speed":{
      "value":877.0,
      "timestamp":"2018-04-12T21:19:49.365873063Z",
      "unit":{
        "symbol":"RPM",
        "name":"revolutions per minute"
      }
    },
    "direction":{
      "value":"forward",
      "timestamp":"2018-04-12T21:19:49.365876827Z",
      "unit":null
    },
    "max_rpm":{
      "value":"1755",
      "timestamp":"2018-04-12T21:19:49.365878629Z",
      "unit":null
    },
    "min_rpm":{
      "value":"175",
      "timestamp":"2018-04-12T21:19:49.365880305Z",
      "unit":null
    }
  }
}
vapor@vec1-c1-wrigley:~$ curl http://${SYNSE_IP}:5000/synse/2.0/fan/chamber/vec/961e1b0ae21c404a0a6a7e47224570ec
{
  "type":"fan",
  "data":{
    "fan_speed":{
      "value":877.0,
      "timestamp":"2018-04-12T21:19:53.296823159Z",
      "unit":{
        "symbol":"RPM",
        "name":"revolutions per minute"
      }
    },
    "direction":{
      "value":"forward",
      "timestamp":"2018-04-12T21:19:53.296828409Z",
      "unit":null
    },
    "max_rpm":{
      "value":"1755",
      "timestamp":"2018-04-12T21:19:53.296831701Z",
      "unit":null
    },
    "min_rpm":{
      "value":"175",
      "timestamp":"2018-04-12T21:19:53.296835005Z",
      "unit":null
    }
  }
}
vapor@vec1-c1-wrigley:~$ curl http://${SYNSE_IP}:5000/synse/2.0/fan/chamber/vec/961e1b0ae21c404a0a6a7e47224570ec?speed_percent=0
[
  {
    "context":{
      "action":"speed_percent",
      "raw":[
        "0"
      ]
    },
    "transaction":"bb7sr09pr5tg00fh8m8g"
  }
]
vapor@vec1-c1-wrigley:~$ curl http://${SYNSE_IP}:5000/synse/2.0/fan/chamber/vec/961e1b0ae21c404a0a6a7e47224570ec
{
  "type":"fan",
  "data":{
    "fan_speed":{
      "value":479.0,
      "timestamp":"2018-04-12T21:20:09.169596778Z",
      "unit":{
        "symbol":"RPM",
        "name":"revolutions per minute"
      }
    },
    "direction":{
      "value":"forward",
      "timestamp":"2018-04-12T21:20:09.169602546Z",
      "unit":null
    },
    "max_rpm":{
      "value":"1755",
      "timestamp":"2018-04-12T21:20:09.169605765Z",
      "unit":null
    },
    "min_rpm":{
      "value":"175",
      "timestamp":"2018-04-12T21:20:09.169608713Z",
      "unit":null
    }
  }
}
vapor@vec1-c1-wrigley:~$ curl http://${SYNSE_IP}:5000/synse/2.0/fan/chamber/vec/961e1b0ae21c404a0a6a7e47224570ec
{
  "type":"fan",
  "data":{
    "fan_speed":{
      "value":348.0,
      "timestamp":"2018-04-12T21:20:11.737539396Z",
      "unit":{
        "symbol":"RPM",
        "name":"revolutions per minute"
      }
    },
    "direction":{
      "value":"forward",
      "timestamp":"2018-04-12T21:20:11.737542972Z",
      "unit":null
    },
    "max_rpm":{
      "value":"1755",
      "timestamp":"2018-04-12T21:20:11.737544935Z",
      "unit":null
    },
    "min_rpm":{
      "value":"175",
      "timestamp":"2018-04-12T21:20:11.737546703Z",
      "unit":null
    }
  }
}
vapor@vec1-c1-wrigley:~$ curl http://${SYNSE_IP}:5000/synse/2.0/fan/chamber/vec/961e1b0ae21c404a0a6a7e47224570ec
{
  "type":"fan",
  "data":{
    "fan_speed":{
      "value":0.0,
      "timestamp":"2018-04-12T21:20:20.95973875Z",
      "unit":{
        "symbol":"RPM",
        "name":"revolutions per minute"
      }
    },
    "direction":{
      "value":"forward",
      "timestamp":"2018-04-12T21:20:20.959743426Z",
      "unit":null
    },
    "max_rpm":{
      "value":"1755",
      "timestamp":"2018-04-12T21:20:20.959745632Z",
      "unit":null
    },
    "min_rpm":{
      "value":"175",
      "timestamp":"2018-04-12T21:20:20.959747448Z",
      "unit":null
    }
  }
}
vapor@vec1-c1-wrigley:~$
```

**Comment by edaniszewski** [comment]
closing this one since a different PR was merged in with this same functionality. will not delete this branch though since there is still stuff in here that should be moved into master, e.g. some cleanup + tests

**Pull request closed without merging**